FEATURE: create a method to anonymize given entities

PRGfx opened the pull request:
Implements the method anonymizeEntity which optionally updates the given
entity in it's repository. For that some internal functionality was
factored out into reusable methods.

PRGfx committed:
FEATURE: create a method to anonymize given entities

Implements the method anonymizeEntity which optionally updates the given
entity in it's repository. For that some internal functionality was
factored out into reusable methods.

theilm commented on Classes/Service/AnonymizationService.php:
```diff
@@ -91,6 +92,22 @@ protected function getAnonymizedPropertyValuesFor($className)
+    protected function isAnonymizable($className)
+    {
+        /** @var AnonymizableEntity $annotation */
+        $annotation = $this->reflectionService->getClassAnnotation($className, AnonymizableEntity::class);
```
I would prefer to use in_array($className, $this->getAnonymizableClassNames) here (and keep getAnonymizableClassNames as is); this prevents using the reflection service twice.

PRGfx replied:
I reverted the changes regarding this method, however this may lead to unfortunate overhead when anonymizing single entities. It is probably advisable that the data currently accumulated to determine anonymizablity in the process is cached in some way.

theilm committed:
Fix typo in README.md

theilm commented on Classes/Service/AnonymizationService.php:
```diff
@@ -99,10 +116,37 @@ protected function getAnonymizedPropertyValuesFor($className)
+                $repository = $this->getRepositoryFor($className);
+                $repository->update($entity);
+            } catch (\Exception $e) {
+                throw new AnonymizationException('Could not determine a repository for the given entity', 1563899697, $e);
```
Please replace the generic exception in the getRepositoryFor with your new AnonymizationException and remove the try/catch here.

PRGfx replied:
Repository::update "might" throw an `IllegalObjectTypeException` (which it probably won't because of the way how we got to the repository in the first place), but is more likely to throw an `UnknownObjectException` in case one tries to update a non-persistent single entity.

I will change the type of the exception but will leave this try/catch statement as is.

_Initially_ the AnonyizationException was created to represent errors closer to the anonymization process, so it may be debatable whether there should be a more persistence related exception type in this package, as even the code part highlighted arguably is a bit off my original intention.

theilm reviewed (changes requested):
Thanks for the feature! What do you think about the proposed changes?

PRGfx committed:
Revert restructuring of isAnonymizable

This may lead to unfortunate overhead when anonymizing single entities.
It is advisable that the data currently accumulated to determine
anonymizablity in the process is cached in some way.

PRGfx committed:
Throw custom exception

Rather than throwing a generic \Exception, a custom exception is thrown
if there is no repository for persisting or retrieving an anonymizable
entity.

PRGfx committed:
TASK bubble persistence exceptions
